### PR TITLE
Eliminate `with mock.patch.dict` for environment variables (3)

### PR DESCRIPTION
**Pull request by bncer**
## Related Issues/PRs

 #9006 

## What changes are proposed in this pull request?

Manual elimination of `with mock.patch.dict` for environment variables and use `monkeypatch.setenv` for mocking. Part 3.

## How is this patch tested?

- [X] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/projects/backend/test_local.py [hunk 1]
```diff
@@ -4,67 +4,75 @@
...
         assert envs == mock_env
 
 
-def test_docker_s3_artifact_cmd_and_envs_from_home():
+def test_docker_s3_artifact_cmd_and_envs_from_home(monkeypatch):
     mock_env = {}
-    with mock.patch.dict("os.environ", mock_env), mock.patch(
-        "posixpath.exists", return_value=True
-    ), mock.patch("posixpath.expanduser", return_value="mock_volume"):
+    monkeypatch.setenvs(mock_env)
```
Comment (gabrielfu) [inline tests/projects/backend/test_local.py:23]: monkeypatch and mock work a bit differently. This line does nothing. You will need to explicitly delete the four env vars:

`AWS_SECRET_ACCESS_KEY`, `AWS_ACCESS_KEY_ID`, `MLFLOW_S3_ENDPOINT_URL`, `MLFLOW_S3_IGNORE_TLS`

File: tests/projects/backend/test_local.py [hunk 2]
```diff
@@ -4,67 +4,75 @@
...
-        assert cmds == ["-v", "/mock_ticket_cache:/mock_ticket_cache"]
-        assert envs == mock_env
+    monkeypatch.delenvs(
+        [
+            "MLFLOW_KERBEROS_TICKET_CACHE",
+            "MLFLOW_KERBEROS_USER",
+            "MLFLOW_PYARROW_EXTRA_CONF",
+        ],
+        raising=False,
+    )
```
Comment (gabrielfu) [inline tests/projects/backend/test_local.py:71]: this is unnecessary

File: tests/store/artifact/test_s3_artifact_repo.py [hunk 3]
```diff
@@ -106,20 +106,23 @@ def test_get_s3_client_hits_cache(s3_artifact_root):
...
-                endpoint_url=ANY,
-                verify=verify,
-                aws_access_key_id=None,
-                aws_secret_access_key=None,
-                aws_session_token=None,
-            )
+def test_get_s3_client_verify_param_set_correctly(
+    s3_artifact_root, ignore_tls_env, verify, monkeypatch
+):
+    monkeypatch.delenv("MLFLOW_S3_IGNORE_TLS", raising=False)
```
Comment (gabrielfu) [inline tests/store/artifact/test_s3_artifact_repo.py:112]: this is unnecessary

**Review by gabrielfu — COMMENTED** [review]
thanks for this PR :) I left a few minor comments

File: tests/projects/backend/test_local.py (hunk 1 above)
Comment (Bncer) [reply tests/projects/backend/test_local.py:23]: Thank you for the review : )
Ok, understand

File: tests/projects/backend/test_local.py (hunk 2 above)
Comment (Bncer) [reply tests/projects/backend/test_local.py:71]: I thought `clear=True` in mock.patch.dict intended to delete envs specified in arguments

Changes were based on my [first commit](https://github.com/mlflow/mlflow/commit/70fab1b608216c9579687534232cabcbbcfe3927#diff-55cacde23edacba590f68eaa71f49ab1591adf3139965bc500d1a62971d3224cL290)  in `test_conda_path()` test:
`with mock.patch.dict("os.environ", mock_env, clear=True):` -> `monkeypatch.delenvs(["CONDA_EXE", MLFLOW_CONDA_HOME], raising=False)`

File: tests/projects/backend/test_local.py [hunk 4]
```diff
@@ -4,67 +4,75 @@
...
+def test_docker_gcs_artifact_cmd_and_envs_from_home(monkeypatch):
     mock_env = {
         "GOOGLE_APPLICATION_CREDENTIALS": "mock_credentials_path",
     }
     gs_uri = "gs://mock_bucket"
-    with mock.patch.dict("os.environ", mock_env, clear=True):
-        cmds, envs = _get_docker_artifact_storage_cmd_and_envs(gs_uri)
-        assert cmds == ["-v", "mock_credentials_path:/.gcs"]
-        assert envs == {"GOOGLE_APPLICATION_CREDENTIALS": "/.gcs"}
+    monkeypatch.delenvs(["GOOGLE_APPLICATION_CREDENTIALS"], raising=False)
```
Comment (Bncer) [inline tests/projects/backend/test_local.py:50]: @gabrielfu The same thing of unnecessary `delenvs` applied here?

Comment (gabrielfu) [reply tests/projects/backend/test_local.py:50]: that's right :)

**Comment by Bncer** [comment]
@gabrielfu please, take a look on fixes from the review

File: tests/projects/backend/test_local.py [hunk 5]
```diff
@@ -4,67 +4,74 @@
...
+            "MLFLOW_S3_IGNORE_TLS",
+        ],
+        raising=False,
+    )
+    with mock.patch("posixpath.exists", return_value=True), mock.patch(
+        "posixpath.expanduser", return_value="mock_volume"
+    ):
         cmds, envs = _get_docker_artifact_storage_cmd_and_envs("s3://mock_bucket")
         assert cmds == ["-v", "mock_volume:/.aws"]
         assert envs == mock_env
```
Comment (gabrielfu) [inline tests/projects/backend/test_local.py:37]: can you remove `mock_env` too?

```suggestion
        assert envs == {}
```

Comment (Bncer) [reply tests/projects/backend/test_local.py]: Done)

**Review by gabrielfu — APPROVED** [review]
thanks @Bncer  :) LGTM!

**Pull request merged**